test/unit/network/ConnectionRegistryTest.js

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const { expect } = require('chai');
+const sinon = require('sinon');
+const sandbox = sinon.createSandbox();
+
+const { ConnectionRegistryImpl } = require('../../../lib/network/ConnectionManager');
+const { Connection } = require('../../../lib/network/Connection');
+const { ConnectionStrategyConfigImpl, ReconnectMode } = require('../../../lib/config/ConnectionStrategyConfig');
+const { RoundRobinLB } = require('../../../lib/util/RoundRobinLB');
+const { UuidUtil } = require('../../../lib/util/UuidUtil');
+const { ClientOfflineError, IOError } = require('../../../lib/core/HazelcastError');
+
+describe('ConnectionRegistryTest', function () {
+
+    const connectionState = {
+        INITIAL: 0,
+        CONNECTED_TO_CLUSTER: 1,
+        INITIALIZED_ON_CLUSTER: 2
+    };
+
+    afterEach(function () {
+        sandbox.restore();
+    });
+
+    describe('getRandomConnection', function () {
+        it('should use load balancer in smart mode', function () {
+            const loadBalancerStub = sandbox.stub(RoundRobinLB.prototype);
+            const connectionRegistry = new ConnectionRegistryImpl(
+                new ConnectionStrategyConfigImpl(),
+                true,
+                loadBalancerStub
+            );
+            connectionRegistry.getRandomConnection();
+            expect(loadBalancerStub.next.called).to.be.true;
+        });
+
+        it('should return first active connection in non-smart mode without using load balancer',
+        function () {
+                const loadBalancerStub = sandbox.stub(RoundRobinLB.prototype);
+                const connectionRegistry = new ConnectionRegistryImpl(
+                    new ConnectionStrategyConfigImpl(),
+                    false,
+                    loadBalancerStub
+                );
+
+                const firstConnection = new sinon.createStubInstance(Connection);
```

Comment:
Are there any advantages in using `new sinon.createStubInstance(Connection)` when compared with `sandbox.stub(Connection.prototype)`? If not, I'd prefer to use `sandbox.stub()` as that's a convention in other tests.

Reply:
Used it since I could not figure out how to create multiple instances from stubbed object. (it says it is wrapped again if stub() is used more than once)

Reply:
I see. BTW do we really need stubs here? Instead we could use `new Connection()` or even `{}`.

Reply:
Harder to construct a connection object, {} should not be what we want here(want to assert right connection is returned)

I think what I used is a good convenience method for multiple stub instances

Reply:
If I can use {}, it is bettee of course

Reply:
In fact `{}` should work just fine:
```js
const firstConnection = {};
connectionRegistry.setConnection(UuidUtil.generate(), firstConnection);                connectionRegistry.setConnection(UuidUtil.generate(), new sinon.createStubInstance(Connection));

const actualConnection = connectionRegistry.getRandomConnection();
expect(actualConnection).to.be.equal(firstConnection);
```

Reply:
Thanks, I will try